

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>How-Tos &mdash; uap 2.0.0 documentation</title>
  

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/graphviz.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Installation of uap" href="installation.html" />
    <link rel="prev" title="Introducing uap" href="introduction.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> uap
          

          
          </a>

          
            
            
              <div class="version">
                2.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introducing <strong>uap</strong></a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#software-design">Software Design</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#recommended-uap-workflow">Recommended <strong>uap</strong> Workflow</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quick Start <strong>uap</strong></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#handle-genomic-data">Handle Genomic Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sequencing-data-analysis">Sequencing Data Analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#general-steps">General Steps</a></li>
<li class="toctree-l3"><a class="reference internal" href="#rnaseq-example-reanalysing-data-from-mercer-link">RNAseq Example – Reanalysing Data from <a href="https://www.ncbi.nlm.nih.gov/pubmed/24705597" target="_blank">Mercer <i>et al.</i>, Nature Protoc. (2014)</a></a></li>
<li class="toctree-l3"><a class="reference internal" href="#chipseq-example-reanalysing-data-from-barski-link">ChIPseq Example – Reanalysing Data from <a href="http://www.ncbi.nlm.nih.gov/pubmed/17512414" target="_blank">Barski <i>et al.</i>, Cell (2007)</a></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#create-your-own-workflow">Create Your Own Workflow</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation of <strong>uap</strong></a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Analysis Configuration File</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html#cluster-configuration-file">Cluster Configuration File</a></li>
<li class="toctree-l1"><a class="reference internal" href="interaction.html">Command-Line Usage of <strong>uap</strong></a></li>
<li class="toctree-l1"><a class="reference internal" href="extension.html">Add New Functionality</a></li>
<li class="toctree-l1"><a class="reference internal" href="platforms.html">Tested Platforms</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation.html">Annotation Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="steps.html">Available steps</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">Information for uap Developers</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">uap</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>How-Tos</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/how-to.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="quick-start-uap">
<span id="how-to"></span><h1>Quick Start <strong>uap</strong><a class="headerlink" href="#quick-start-uap" title="Permalink to this headline">¶</a></h1>
<p>At first, you need to install <strong>uap</strong> (see <a class="reference internal" href="installation.html"><span class="doc">Installation of uap</span></a>).
After successfully finishing the installation of <strong>uap</strong> example
analysis can be found in the folder <code class="docutils literal notranslate"><span class="pre">example-configurations</span></code>.</p>
<p>Let’s jump head first into <strong>uap</strong> and have a look at some examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd &lt;uap-path&gt;/example-configurations/
$ ls *.yaml
2007-CD4+_T_Cell_ChIPseq-Barski_et_al_download.yaml
2007-CD4+_T_Cell_ChIPseq-Barski_et_al.yaml
2014-RNA_CaptureSeq-Mercer_et_al_download.yaml
2014-RNA_CaptureSeq-Mercer_et_al.yaml
download_human_gencode_release.yaml
index_homo_sapiens_hg19_genome.yaml
index_mycoplasma_genitalium_ASM2732v1_genome.yaml
</pre></div>
</div>
<p>These example configurations differ in their usage of computational
resources.
Some example configurations download or work on small datasets and are
thus feasible for machines with limited resources.
Most examples can be extended by uncommenting additional steps.
This might change their computational requirements in such a way that a
very powerful stand-alone machine or a cluster system is required.
The examples are marked accordingly in the sections below.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Before <strong>computing an example on a cluster</strong>, you need to uncomment
the <a class="reference internal" href="configuration.html#config-file-cluster"><span class="std std-ref">cluster Section</span></a> and adapt the settings as required.
Please check also if the <a class="reference internal" href="configuration.html#cluster-configuration"><span class="std std-ref">Cluster Configuration File</span></a> fits your
cluster system.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The examples contain information where users can obtain
<strong>required external/bioinformatics tools</strong>.
If <strong>uap</strong> fails due to a missing tool, please check the
provided URLs for installation instructions.</p>
</div>
<div class="section" id="handle-genomic-data">
<h2>Handle Genomic Data<a class="headerlink" href="#handle-genomic-data" title="Permalink to this headline">¶</a></h2>
<p>A usual analysis of High-Throughput Sequencing (HTS) data relies on different
publicly available data.
Most important is probably the genomic sequence of the species under
investigation.
That sequence is required to construct the indices (data structures used by
read aligners).
Other publicly available data sets (such as reference annotations or the
chromosome sizes) might also be required for an analysis.
The following configurations showcase how to get or generate that data:</p>
<dl id="example-index-mycoplasma">
<dt><code class="docutils literal notranslate"><span class="pre">index_mycoplasma_genitalium_ASM2732v1_genome.yaml</span></code></dt><dd><p>Downloads the <em>Mycoplasma genitalium</em> genome, generates the indices for
<a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">bowtie2</a>, <a href="http://bio-bwa.sourceforge.net/" target="_blank">bwa</a>, <a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a>, and <a href="http://www.htslib.org/" target="_blank">samtools</a>.
This workflow is quite fast because it uses the very small genome of
<em>Mycoplasma genitalium</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Max. memory</dt>
<dd class="field-odd"><p>~0,5 GB</p>
</dd>
<dt class="field-even">Disk usage</dt>
<dd class="field-even"><p>~20 MB</p>
</dd>
<dt class="field-odd">Run time</dt>
<dd class="field-odd"><p>minutes</p>
</dd>
</dl>
<p>Required tools:</p>
<blockquote>
<div><ul class="simple">
<li><p><a href="http://bio-bwa.sourceforge.net/" target="_blank">bwa</a></p></li>
<li><p><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">bowtie2</a></p></li>
<li><p><a href="https://curl.haxx.se/" target="_blank">curl</a></p></li>
<li><p><a href="http://zlib.net/pigz/" target=_blank">pigz</a></p></li>
<li><p><a href="http://www.htslib.org/" target="_blank">samtools</a></p></li>
<li><p><a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a></p></li>
</ul>
</div></blockquote>
</dd>
</dl>
<dl id="example-index-hg19">
<dt><code class="docutils literal notranslate"><span class="pre">index_homo_sapiens_hg19_genome.yaml</span></code></dt><dd><p>Downloads chromosome 21 of the <em>Homo sapiens</em> genome, generates the indices
for <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">bowtie2</a>, <a href="http://bio-bwa.sourceforge.net/" target="_blank">bwa</a>, and <a href="http://www.htslib.org/" target="_blank">samtools</a>.
This minimal version should work just fine.
Users can uncomment steps to download the complete genome.
This would substantially increase the required computational resources.
The <a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a> index creation is commented out due to its high
memory consumption (~50-60 GB), if working with the whole genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Max. memory</dt>
<dd class="field-odd"><p>~2 GB</p>
</dd>
<dt class="field-even">Disk usage</dt>
<dd class="field-even"><p>~240 MB</p>
</dd>
<dt class="field-odd">Run time</dt>
<dd class="field-odd"><p>several minutes</p>
</dd>
</dl>
<p>Downloads <em>Homo sapiens</em> chromosome 21, generates the indices for
<a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">bowtie2</a>, <a href="http://bio-bwa.sourceforge.net/" target="_blank">bwa</a>, and <a href="http://www.htslib.org/" target="_blank">samtools</a> (and <a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a> if
you uncomment it).
This workflow requires substantial computational resources due to the
size of the human genome.
The <a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a> index creation is commented out due to its high
memory consumption.
Please make sure to only run it on well equipped machines.</p>
<p>Required tools:</p>
<blockquote>
<div><ul class="simple">
<li><p><a href="http://bio-bwa.sourceforge.net/" target="_blank">bwa</a></p></li>
<li><p><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">bowtie2</a></p></li>
<li><p><a href="https://curl.haxx.se/" target="_blank">curl</a></p></li>
<li><p><a href="http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/fetchChromSizes" target="_blank">fetchChromSizes</a></p></li>
<li><p><a href="http://zlib.net/pigz/" target=_blank">pigz</a></p></li>
<li><p><a href="http://www.htslib.org/" target="_blank">samtools</a></p></li>
<li><p><a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a> (if uncommented)</p></li>
</ul>
</div></blockquote>
</dd>
</dl>
<dl id="example-download-gencode">
<dt><code class="docutils literal notranslate"><span class="pre">download_human_gencode_release.yaml</span></code></dt><dd><p>Downloads the human Gencode main annotation v19 and a subset for long
non-coding RNA genes.
This workflow only downloads files from the internet and and thus should
work on any machine.</p>
<dl class="field-list simple">
<dt class="field-odd">Max. memory</dt>
<dd class="field-odd"><p>depends on your machine</p>
</dd>
<dt class="field-even">Disk usage</dt>
<dd class="field-even"><p>~1,2 GB</p>
</dd>
<dt class="field-odd">Run time</dt>
<dd class="field-odd"><p>depends on your internet connection</p>
</dd>
</dl>
<p>Required tools:</p>
<blockquote>
<div><ul class="simple">
<li><p><a href="http://bio-bwa.sourceforge.net/" target="_blank">bwa</a></p></li>
<li><p><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">bowtie2</a></p></li>
<li><p><a href="https://curl.haxx.se/" target="_blank">curl</a></p></li>
<li><p><a href="http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/fetchChromSizes" target="_blank">fetchChromSizes</a></p></li>
<li><p><a href="http://zlib.net/pigz/" target=_blank">pigz</a></p></li>
<li><p><a href="http://www.htslib.org/" target="_blank">samtools</a></p></li>
</ul>
</div></blockquote>
</dd>
</dl>
<p>Let’s have a look at the <em>Mycoplasma genitalium</em> example workflow by checking
its <span class="xref std std-ref">uap_status</span>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd &lt;uap-path&gt;/example-configurations/
$ uap index_mycoplasma_genitalium_ASM2732v1_genome.yaml status
[uap][ERROR]: index_mycoplasma_genitalium_ASM2732v1_genome.yaml: Destination path does not exist: genomes/bacteria/Mycoplasma_genitalium/
</pre></div>
</div>
<p>Oops, the <code class="docutils literal notranslate"><span class="pre">destination_path</span></code> does not exist (see <a class="reference internal" href="configuration.html#config-file-destination-path"><span class="std std-ref">destination_path Section</span></a>).
Create it and start again:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ mkdir -p genomes/bacteria/Mycoplasma_genitalium/
$ uap index_mycoplasma_genitalium_ASM2732v1_genome.yaml status

Waiting tasks
-------------
[w] bowtie2_index/Mycoplasma_genitalium_index-download
[w] bwa_index/Mycoplasma_genitalium_index-download
[w] fasta_index/download
[w] segemehl_index/Mycoplasma_genitalium_genome-download

Ready tasks
-----------
[r] M_genitalium_genome/download

tasks: 5 total, 4 waiting, 1 ready
</pre></div>
</div>
<p>A list with all runs and their respective state should be displayed.
A run is always in one of these states:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">[r]eady</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">[w]aiting</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">[q]ueued</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">[e]xecuting</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">[f]inished</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">[c]hanged</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">[b]ad</span></code></p></li>
</ul>
<p>If the command still fails, please check that the tools defined in
<code class="docutils literal notranslate"><span class="pre">index_mycoplasma_genitalium_ASM2732v1_genome.yaml</span></code> are available in your
environment (see <a class="reference internal" href="configuration.html#config-file-tools"><span class="std std-ref">tools Section</span></a>).
If you really want to download and index the genome tell <strong>uap</strong> to start
the workflow:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ uap index_mycoplasma_genitalium_ASM2732v1_genome.yaml run-locally
</pre></div>
</div>
<p><strong>uap</strong> should have created a symbolic link named
<code class="docutils literal notranslate"><span class="pre">index_mycoplasma_genitalium_ASM2732v1_genome.yaml-out</span></code> pointing to the
<code class="docutils literal notranslate"><span class="pre">destination_path</span></code>.
The content should look something like that:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ tree --charset=ascii
.
|-- bowtie2_index
|   |-- Mycoplasma_genitalium_index-download-cMQPtBxs
|   |   |-- Mycoplasma_genitalium_index-download.1.bt2
|   |   |-- Mycoplasma_genitalium_index-download.2.bt2
|   |   |-- Mycoplasma_genitalium_index-download.3.bt2
|   |   |-- Mycoplasma_genitalium_index-download.4.bt2
|   |   |-- Mycoplasma_genitalium_index-download.rev.1.bt2
|   |   `-- Mycoplasma_genitalium_index-download.rev.2.bt2
|   `-- Mycoplasma_genitalium_index-download-ZsvbSjtK
|       |-- Mycoplasma_genitalium_index-download.1.bt2
|       |-- Mycoplasma_genitalium_index-download.2.bt2
|       |-- Mycoplasma_genitalium_index-download.3.bt2
|       |-- Mycoplasma_genitalium_index-download.4.bt2
|       |-- Mycoplasma_genitalium_index-download.rev.1.bt2
|       `-- Mycoplasma_genitalium_index-download.rev.2.bt2
|-- bwa_index
|   `-- Mycoplasma_genitalium_index-download-XRyj5AnJ
|       |-- Mycoplasma_genitalium_index-download.amb
|       |-- Mycoplasma_genitalium_index-download.ann
|       |-- Mycoplasma_genitalium_index-download.bwt
|       |-- Mycoplasma_genitalium_index-download.pac
|       `-- Mycoplasma_genitalium_index-download.sa
|-- fasta_index
|   `-- download-HA439DGO
|       `-- Mycoplasma_genitalium.ASM2732v1.fa.fai
|-- M_genitalium_genome
|   `-- download-5dych7Xj
|-- Mycoplasma_genitalium.ASM2732v1.fa
|-- segemehl_index
|   |-- Mycoplasma_genitalium_genome-download-2UKxxupJ
|   |   |-- download-segemehl-generate-index-log.txt
|   |   `-- Mycoplasma_genitalium_genome-download.idx
|   `-- Mycoplasma_genitalium_genome-download-zgtEpQmV
|       |-- download-segemehl-generate-index-log.txt
|       `-- Mycoplasma_genitalium_genome-download.idx
`-- temp
</pre></div>
</div>
<p>Congratulation you’ve finished your first <strong>uap</strong> workflow!</p>
<p>Go on and try to run some more workflows.
Most examples require the human genome so you might turn your head towards the
<code class="docutils literal notranslate"><span class="pre">index_homo_sapiens_hg19_genome.yaml</span></code> workflow from her:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ uap index_homo_sapiens_hg19_genome.yaml status
[uap][ERROR]: Output directory (genomes/animalia/chordata/mammalia/primates/homo_sapiens/hg19/chromosome_sizes) does not exist. Please create it.
$ mkdir -p genomes/animalia/chordata/mammalia/primates/homo_sapiens/hg19/chromosome_sizes
$ uap index_homo_sapiens_hg19_genome.yaml run-locally
&lt;Analysis starts&gt;
</pre></div>
</div>
<p>Again you need to create the output folder (you get the idea).
Be aware that by default only the smallest chromosome, chromsome 21, is
downloaded and indexed.
This reduces required memory and computation time.
You can uncomment the download steps for the other chromosomes and the index
for the complete genome will be created.</p>
</div>
<div class="section" id="sequencing-data-analysis">
<h2>Sequencing Data Analysis<a class="headerlink" href="#sequencing-data-analysis" title="Permalink to this headline">¶</a></h2>
<p>Now that you possess the genome sequences, indices, and annotations let’s have
a look at some example analysis.</p>
<div class="section" id="general-steps">
<h3>General Steps<a class="headerlink" href="#general-steps" title="Permalink to this headline">¶</a></h3>
<p>The analysis of high-throughput sequencing (HTS) data usually start with some
basic steps.</p>
<ol class="arabic simple">
<li><p>Conversion of the raw sequencing data to, most likely, fastq(.gz) files</p></li>
<li><p>Removal of adapter sequences from the sequencing reads</p></li>
<li><p>Alignment of the sequencing reads onto the reference genome</p></li>
</ol>
<p>These basic steps can be followed up with a lot of different analysis steps.
The following analysis examples illustrate how to perform the basic as well as
some more specific steps.</p>
</div>
<div class="section" id="rnaseq-example-reanalysing-data-from-mercer-link">
<h3>RNAseq Example – Reanalysing Data from <a href="https://www.ncbi.nlm.nih.gov/pubmed/24705597" target="_blank">Mercer <i>et al.</i>, Nature Protoc. (2014)</a><a class="headerlink" href="#rnaseq-example-reanalysing-data-from-mercer-link" title="Permalink to this headline">¶</a></h3>
<p>RNAseq analysis often aims at the discovery of differentially expressed
(known) transcripts. Therefore mappped reads for at least two different samples
have to be available.</p>
<ol class="upperalpha simple">
<li><p>Differential Expression Analysis</p>
<ol class="arabic simple" start="4">
<li><p>Get annotation set (for e.g. genes, transcripts, …)</p></li>
<li><p>Count the number of reads overlapping the annotation</p></li>
<li><p>Perform statistical analysis, based on counts</p></li>
</ol>
</li>
</ol>
<p>Another common analysis performed with RNAseq data is the identification of
novel tarnscripts. This approach is useful to identify tissue-specific
transcipts.</p>
<ol class="upperalpha simple" start="2">
<li><p><em>De novo</em> Transcript Assembly</p>
<ol class="arabic simple" start="4">
<li><p>Apply transcript assembly tool on mapped reads</p></li>
</ol>
</li>
</ol>
<dl id="example-mercer-download">
<dt><code class="docutils literal notranslate"><span class="pre">2014-RNA_CaptureSeq-Mercer_et_al_download.yaml</span></code></dt><dd><p>Downloads the data published in the paper <a href="https://www.ncbi.nlm.nih.gov/pubmed/24705597" target="_blank">Mercer <i>et al.</i>, Nature Protoc. (2014)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Max. memory</dt>
<dd class="field-odd"><p>~? GB</p>
</dd>
<dt class="field-even">Disk usage</dt>
<dd class="field-even"><p>~12 GB</p>
</dd>
<dt class="field-odd">Run time</dt>
<dd class="field-odd"><p>minutes (depending on your internet connection)</p>
</dd>
</dl>
<p>Required tools:</p>
<blockquote>
<div><ul class="simple">
<li><p><a href="https://curl.haxx.se/" target="_blank">curl</a></p></li>
<li><p><a href="http://zlib.net/pigz/" target=_blank">pigz</a></p></li>
</ul>
</div></blockquote>
</dd>
</dl>
<dl class="simple" id="example-mercer">
<dt><code class="docutils literal notranslate"><span class="pre">2014-RNA_CaptureSeq-Mercer_et_al.yaml</span></code></dt><dd><p>The downloaded FASTQ files get analysed by <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/" target="_blank">FastQC</a> and
<a href="http://hannonlab.cshl.edu/fastx_toolkit/" target="_blank">FASTX-Toolkit</a>.
The reads are afterwards mapped to the human genome with <a href="https://ccb.jhu.edu/software/tophat/index.shtml" target="_blank">tophat2</a>.
The mapped reads are afterwards sorted by position using <a href="http://www.htslib.org/" target="_blank">samtools</a>.
<a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html" target="_blank">htseq-count</a> is used to count the mapped reads for every exon of
the annotation.
<a href="http://cufflinks.cbcb.umd.edu/" target="_blank">cufflinks</a> is used to perform <em>de novo</em> transcript assembly.
The usage of <a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a> is <strong>disabled</strong> by default.
But it can be enabled and combined with <a href="http://cufflinks.cbcb.umd.edu/" target="_blank">cufflinks</a> <em>de novo</em>
transcript assembly employing our <strong>s2c</strong> python script.</p>
<dl class="field-list simple">
<dt class="field-odd">Max. memory</dt>
<dd class="field-odd"><p>~? GB</p>
</dd>
<dt class="field-even">Disk usage</dt>
<dd class="field-even"><p>~3 GB</p>
</dd>
<dt class="field-odd">Run time</dt>
<dd class="field-odd"><p>several hours</p>
<ul class="simple">
<li><p><a href="http://cufflinks.cbcb.umd.edu/" target="_blank">cufflinks</a></p></li>
<li><p><a href="https://github.com/marcelm/cutadapt" target="_blank">cutadapt</a></p></li>
<li><p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/" target="_blank">FastQC</a></p></li>
<li><p><a href="http://hannonlab.cshl.edu/fastx_toolkit/" target="_blank">FASTX-Toolkit</a></p></li>
<li><p><a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html" target="_blank">htseq-count</a></p></li>
<li><p><a href="http://zlib.net/pigz/" target=_blank">pigz</a></p></li>
<li><p><a href="http://www.htslib.org/" target="_blank">samtools</a></p></li>
<li><p><a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a> (if uncommented)</p></li>
<li><p><a href="https://ccb.jhu.edu/software/tophat/index.shtml" target="_blank">tophat2</a></p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Before computing <code class="docutils literal notranslate"><span class="pre">2014-RNA_CaptureSeq-Mercer_et_al.yaml</span></code>
please make sure that, the following examples were executed:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">index_homo_sapiens_hg19_genome.yaml</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">download_human_gencode_release.yaml</span></code></p></li>
</ul>
</div>
</div>
<div class="section" id="chipseq-example-reanalysing-data-from-barski-link">
<h3>ChIPseq Example – Reanalysing Data from <a href="http://www.ncbi.nlm.nih.gov/pubmed/17512414" target="_blank">Barski <i>et al.</i>, Cell (2007)</a><a class="headerlink" href="#chipseq-example-reanalysing-data-from-barski-link" title="Permalink to this headline">¶</a></h3>
<p>ChIPseq analysis aims at the discovery of genomic loci at which protein(s) of
interest were bound. The experiment is an enrichment procedure using specific
antibodies. The enrichment detection is normally performed by so called peak
calling programs. The data is prone to duplicate reads from PCR due to relatively
low amounts of input DNA. So these steps follow the basic ones:</p>
<ol class="arabic simple" start="4">
<li><p>Duplicate removal</p></li>
<li><p>Peak calling</p></li>
</ol>
<p>The analysis of data published in the paper <a href="http://www.ncbi.nlm.nih.gov/pubmed/17512414" target="_blank">Barski <i>et al.</i>, Cell (2007)</a> is contained in these
files:</p>
<dl id="example-barski-download">
<dt><code class="docutils literal notranslate"><span class="pre">2007-CD4+_T_Cell_ChIPseq-Barski_et_al_download.yaml</span></code></dt><dd><p>Downloads the data published in the paper <a href="http://www.ncbi.nlm.nih.gov/pubmed/17512414" target="_blank">Barski <i>et al.</i>, Cell (2007)</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Max. memory</dt>
<dd class="field-odd"><p>~? GB</p>
</dd>
<dt class="field-even">Disk usage</dt>
<dd class="field-even"><p>~17 GB</p>
</dd>
<dt class="field-odd">Run time</dt>
<dd class="field-odd"><p>depends on your internet connection</p>
</dd>
</dl>
<p>Downloads the data published in the paper <a href="http://www.ncbi.nlm.nih.gov/pubmed/17512414" target="_blank">Barski <i>et al.</i>, Cell (2007)</a>.</p>
<p>Required tools:</p>
<blockquote>
<div><ul class="simple">
<li><p><a href="https://curl.haxx.se/" target="_blank">curl</a></p></li>
<li><p><a href="http://zlib.net/pigz/" target=_blank">pigz</a></p></li>
</ul>
</div></blockquote>
</dd>
</dl>
<dl id="example-barski">
<dt><code class="docutils literal notranslate"><span class="pre">2007-CD4+_T_Cell_ChIPseq-Barski_et_al.yaml</span></code></dt><dd><p>At first the downloaded FASTQ files are grouped by sample.
All files per sample are merged.
Sequencing quality is controlled by <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/" target="_blank">FastQC</a> and <a href="http://hannonlab.cshl.edu/fastx_toolkit/" target="_blank">FASTX-Toolkit</a>.
Adapter sequences are removed from the reads before they are mapped to
the human genome.
Reads are mapped with <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">bowtie2</a>, <a href="http://bio-bwa.sourceforge.net/" target="_blank">bwa</a>, and <a href="https://ccb.jhu.edu/software/tophat/index.shtml" target="_blank">tophat2</a>.
Again mapping with <a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a> is disabled by default due to its
high resource requirements.
Library complexity is estimated using <a href="http://smithlabresearch.org/software/preseq/" target="_blank">preseq</a>.
After the mapping duplicate reads are removed using <a href="http://broadinstitute.github.io/picard/" target="_blank">Picard</a>.
Finally enriched regions are detected with <a href="https://github.com/taoliu/MACS" target="_blank">MACS2</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Max. memory</dt>
<dd class="field-odd"><p>~? GB</p>
</dd>
<dt class="field-even">Disk usage</dt>
<dd class="field-even"><p>~51 GB</p>
</dd>
<dt class="field-odd">Run time</dt>
<dd class="field-odd"><p>~several hours (on a cluster), ~1 day (on a single machine)</p>
</dd>
</dl>
<p>Required tools:</p>
<blockquote>
<div><ul class="simple">
<li><p><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">bowtie2</a></p></li>
<li><p><a href="http://bio-bwa.sourceforge.net/" target="_blank">bwa</a></p></li>
<li><p><a href="https://github.com/marcelm/cutadapt" target="_blank">cutadapt</a></p></li>
<li><p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/" target="_blank">FastQC</a></p></li>
<li><p><a href="http://hannonlab.cshl.edu/fastx_toolkit/" target="_blank">FASTX-Toolkit</a></p></li>
<li><p><a href="https://github.com/taoliu/MACS" target="_blank">MACS2</a></p></li>
<li><p><a href="http://broadinstitute.github.io/picard/" target="_blank">Picard</a></p></li>
<li><p><a href="http://zlib.net/pigz/" target=_blank">pigz</a></p></li>
<li><p><a href="http://smithlabresearch.org/software/preseq/" target="_blank">preseq</a></p></li>
<li><p><a href="http://www.htslib.org/" target="_blank">samtools</a></p></li>
<li><p><a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/" target="_blank">segemehl</a></p></li>
<li><p><a href="https://ccb.jhu.edu/software/tophat/index.shtml" target="_blank">tophat2</a></p></li>
</ul>
</div></blockquote>
</dd>
</dl>
<div class="admonition hint">
<p class="admonition-title">Hint</p>
<p>The usage of <a href="http://broadinstitute.github.io/picard/" target="_blank">Picard</a> can differ a lot between systems.
On Ubuntu systems it can be called like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ picard-tools --version
</pre></div>
</div>
<p>If you use it as recommended at <a href="http://broadinstitute.github.io/picard/" target="_blank">Picard</a>, it is called like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ java -jar /path/to/picard.jar -h
</pre></div>
</div>
<p>Please check how to use it on your system and adjust the example
configuration accordingly (see <a class="reference internal" href="configuration.html#config-file-tools"><span class="std std-ref">tools Section</span></a>).</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Before computing <code class="docutils literal notranslate"><span class="pre">2007-CD4+_T_Cell_ChIPseq-Barski_et_al.yaml</span></code>
please make sure that, the following examples were executed:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">index_homo_sapiens_hg19_genome.yaml</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">download_human_gencode_release.yaml</span></code></p></li>
</ul>
</div>
<div class="section" id="create-your-own-workflow">
<h4>Create Your Own Workflow<a class="headerlink" href="#create-your-own-workflow" title="Permalink to this headline">¶</a></h4>
<p>You finished to check out the examples?
Go and try to create your own workflow
If you are fine with what you saw
Although writing the configuration may seem a bit complicated, the trouble
pays off later because further interaction with the pipeline is quite simple.
The structure and content of the configuration files is very detailed described
on another page (see <a class="reference internal" href="configuration.html#analysis-configuration"><span class="std std-ref">Analysis Configuration File</span></a>).</p>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="installation.html" class="btn btn-neutral float-right" title="Installation of uap" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="introduction.html" class="btn btn-neutral float-left" title="Introducing uap" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, Christoph Kämpf, Michael Specht

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>